controllers/awsmachine_controller.go

```diff
@@ -911,3 +913,19 @@ func (r *AWSMachineReconciler) indexAWSMachineByInstanceID(o client.Object) []st
 
 	return nil
 }
+
+func (r *AWSMachineReconciler) ensureStorageTags(ec2svc services.EC2MachineInterface, instanceID *string, machine *infrav1.AWSMachine) {
+	volumes, err := ec2svc.GetEC2Volumes(instanceID)
```

Comment:
I think this is too much overhead in terms of api calls as we get the volumes first and then check if there is a change in the tags. So GetEC2Volumes is called at each reconcile.

Let's think about a way to avoid it.
We get the volumeId info from `v.BlockDeviceMappings[0].Ebs.VolumeId` in this method:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/c53c107cbddb9e32e79455b1b7f96d46a7aafb08/pkg/cloud/services/ec2/instances.go#L761

We have annotations on the AWSMachine, we have additionaltags. If there is no change, we don't call GetEC2Volumes(). If there is change, we still don't call GetEC2Volumes(), just update the tags with UpdateResourceTags() using the volumeId and then update the annotation.

Does this sound reasonable?

Reply:
sounds fair, but again to get this v *ec2.instance, we need to call describeInstances which is again an API call, is my understanding correct?

Reply:
We will have already have called describeInstances by the time we have got here, so the suggestion is to change the `ensureStorageTags` signature to pass in a infrav1.Instance .

@sedefsavas  is suggesting that if we know from the annotation that there is no change in tags to be done, we do not need to call GetEC2Volumes, which will minimise calls to the EC2 API.

Reply:
As per the comment --> If there is change, we still don't call GetEC2Volumes(), just update the tags with UpdateResourceTags() using the volumeId and then update the annotation ---> i understood from this statement that we shouldn't use getEC2Volumes at all, and make use of volume id from` v.BlockDeviceMappings[0].Ebs.VolumeId`